File: spring-boot-3-demo-app/src/test/java/io/github/mfvanek/spring/boot3/test/support/JaegerInitializer.java
package io.github.mfvanek.spring.boot3.test.support;

import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.utility.DockerImageName;

import javax.annotation.Nonnull;

@SuppressWarnings("resource")
public class JaegerInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

    private static final DockerImageName IMAGE = DockerImageName.parse("jaegertracing/all-in-one:1.53");
    private static final GenericContainer<?> JAEGER = new GenericContainer<>(IMAGE)
        .withExposedPorts(4317);

    @Override
    public void initialize(final ConfigurableApplicationContext context) {
        JAEGER.start();

        final String jaegerUrl = "http://localhost:" + JAEGER.getFirstMappedPort();
        TestPropertyValues.of(
            "management.otlp.tracing.endpoint=" + jaegerUrl
        ).applyTo(context.getEnvironment());
    }

    @Nonnull
    public static Integer getFirstMappedPort() {
        return JAEGER.getFirstMappedPort();
    }
}

File: spring-boot-3-demo-app-kotlin/src/test/kotlin/io/github/mfvanek/spring/boot3/kotlin/test/support/JaegerInitializer.kt
package io.github.mfvanek.spring.boot3.kotlin.test.support

import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.testcontainers.containers.GenericContainer
import org.testcontainers.utility.DockerImageName

class JaegerInitializer : ApplicationContextInitializer<ConfigurableApplicationContext> {

    override fun initialize(context: ConfigurableApplicationContext) {
        JAEGER.start()
        val jaegerUrl = "http://localhost:" + JAEGER.firstMappedPort
        TestPropertyValues.of(
            "management.otlp.tracing.endpoint=$jaegerUrl"
        ).applyTo(context.environment)
    }

    companion object {
        @JvmStatic
        private val IMAGE: DockerImageName = DockerImageName.parse("jaegertracing/all-in-one:1.53")

        @JvmStatic
        private val JAEGER = GenericContainer(IMAGE).withExposedPorts(4317)

        @JvmStatic
        fun getFirstMappedPort(): Int = JAEGER.firstMappedPort
    }
}
